### PR TITLE
feat: turn invalid numbers into strings

### LINKED ISSUE #120: Improve repairing strings missing quotes like `0.0.1` and `ES2020`
Hi @josdejong, thanks a lot for such a great library, really helps me out.

I have found that jsonrepair will fail in some cases, I wonder if this is expected?

1. A non-number appears after a number
image
image
2. Any character that appears before/after boolean
image
image
3. `-+`
image
image

I want to open a PR to fix these, can you confirm which ones need to be fixed? thx @josdejong

### PR DESCRIPTION
**Pull request by keyserj**
Fixes #120 

The changes here are within the number parsing logic:

- after parsing a numeric symbol (.-eE), if we're not at the end and the next char is not a digit, instead of throwing, fall back to trying to parse other types (in these cases, unquoted string).
- if we've already parsed a numeric symbol once and we've parsed through all the digits following that, instead of taking what's been parsed and assuming trailing alphanumerics are ok, fall back to trying to parse other types (again in this case, unquoted string).

Tests currently pass for regular but fail for streams (in parseRootEnd: Error: Unexpected character "." at position 3). Seems like it might just be an issue with properly updating the stack which I'm having a hard time figuring out. I could spend more time trying to understand but it seems worth asking - does anyone know what could be wrong with the code for streams here?

P.S.: thanks for the library! and the non-stream code was easy to follow 🔥

### CODE REVIEW

File: src/index.test.ts [hunk 1]
```diff
@@ -441,6 +441,18 @@ describe.each(implementations)('jsonrepair [$name]', ({ jsonrepair }) => {
...
 
+    test('should turn invalid numbers into strings', () => {
+      expect(jsonrepair('ES2020')).toBe('"ES2020"')
+      // TODO: most of these still failing in parseRootEnd for streams: Error: Unexpected character "." at position 3
+      expect(jsonrepair('0.0.1')).toBe('"0.0.1"')
+      expect(jsonrepair('746de9ad-d4ff-4c66-97d7-00a92ad46967')).toBe('"746de9ad-d4ff-4c66-97d7-00a92ad46967"')
+      expect(jsonrepair('234..5')).toBe('"234..5"')
+      expect(jsonrepair('[0.0.1,2]')).toBe('["0.0.1",2]') // test delimiter for numerics
+      expect(jsonrepair('[2 0.0.1 2]')).toBe('[2, "0.0.1 2"]') // note: currently spaces delimit numbers, but don't delimit unquoted strings
+      expect(jsonrepair('2.3.4')).toBe('"2.3.4"')
```
Comment (keyserj) [inline src/index.test.ts:452]: looking at this again, I had moved it from the "expect to throw" tests, but it's testing the same thing at the `0.0.1 to be "0.0.1"` test, so I'll remove it

**Comment by josdejong** [comment]
Thanks for your PR Joel, I'll look into it this week. It looks promising!

> P.S.: thanks for the library! and the non-stream code was easy to follow 🔥

Thanks, good to hear 😊

File: src/index.test.ts (hunk 1 above)
Comment (josdejong) [reply src/index.test.ts:452]: Yes, perfect!

**Comment by josdejong** [comment]
This looks really solid!

It is unfortunate that the same pattern of code is repeated three times, but I don't see a handy way to reuse that code since it does not always return something. So let's just keep it as you've implemented it right now.

> Tests currently pass for regular but fail for streams (in parseRootEnd: Error: Unexpected character "." at position 3)

It turns out you forgot to add the following piece of code to the streaming version, close to the end of the `parseNumber` function:

```js
// if we're not at the end of the number by this point, allow this to be parsed as another type
if (!atEndOfNumber()) {
  i = start
  return false
}
```
After adding that, all tests turn green 🎉

**Comment by josdejong** [comment]
On a side note: I see that the CI tests in Github Actions where not running for PR's, should be fixed via https://github.com/josdejong/jsonrepair/commit/eacf5cefe8b18c4305023cac9a4ee847c0de9b8d

**Comment by keyserj** [comment]
> It turns out you forgot to add the following piece of code to the streaming version

🤦 😄 thanks for finding that! Glad it was something simple.

> It is unfortunate that the same pattern of code is repeated three times, but I don't see a handy way to reuse that code since it does not always return something.

Yeah I was thinking the same thing. Potentially we could go for something like this:

```ts
  type ParsingStatus = 'complete' | 'invalid' | 'incomplete'

  function handleNumericSymbol(start: number): ParsingStatus {
    if (atEndOfNumber()) {
      repairNumberEndingWithNumericSymbol(start)
      return 'complete'
    }
    if (!isDigit(text.charCodeAt(i))) {
      i = start
      return 'invalid'
    }
    while (isDigit(text.charCodeAt(i))) {
      i++
    }
    return 'incomplete'
  }

  ...

  function parseNumber(): boolean {
    const start = i
    if (text.charCodeAt(i) === codeMinus) {
      i++
      const status = handleNumericSymbol(start)
      if (status === 'complete') return true
      if (status === 'invalid') return false
    }
    ...
  }
```

Which has the trade off of still repeating some code (status checks) but not repeating the main logic of it. I'll leave this change out for now but let me know if you think it's worth adding.

**Comment by josdejong** [comment]
Thanks for the update!

We can indeed introduce a function that returns three different states, but indeed at the place where you need it you have to add extra logic then. All in all it adds complexity rather than reducing it, so I think it's best to leave the repetitive code as it is for now. I guess what we would need there is a macro :). Anyway, let's not overengineer this. 

I'm really happy with this new capability for jsonrepair, I think it is quite a common case, thanks for picking this up!

**Comment by josdejong** [comment]
Published now in `v3.7.0`.

**Pull request merged**